Add gateway-api reference architecture (#167)
canonical/ingress-configurator-operator: docs/explanation/gateway-route.md

    Workload -- "ingress relation" --> IC
    IC -- "gateway-route relation" --> GAI

    classDef k8s stroke:#2e7d32
    classDef juju stroke:#1565c0,stroke-dasharray:5 5
    linkStyle 0,1,2,3 stroke:#2e7d32
    linkStyle 4,5 stroke:#1565c0,stroke-dasharray:5 5
```

## How HTTPRoute resources are created

The Ingress Configurator builds `HTTPRoute` resources based on the
`https_mode` received from the provider (see
{ref}`The gateway-route relation <reference_gateway_route>`) and the
hostname(s) from its own configuration:

- **One HTTP route** is always created. It covers all hostnames and attaches
  to every per-hostname HTTP listener on the `Gateway` via multiple
  `parentRefs`. When no hostnames are configured, it falls back to the
  `Gateway`'s hostname-less HTTP listener.
- When `https_mode` is `enabled` or `enforced`, **one HTTPS route per
  hostname** is created, each attaching to its corresponding per-hostname HTTPS
  listener.
- When `https_mode` is `enforced`, the HTTP route does **not** forward to
  the backend. Instead it issues a `301` HTTPS redirect using a
  `RequestRedirect` filter.

(explanation_backend_service_selection)=

## Backend service selection

> The `HTTPRoute` forwards traffic to a Kubernetes `Service`. Which `Service` is

Reviewer: It feels a bit redundant with the "Architecture" intro, is it on purpose?

used depends on whether the workload charm has opened its port in Juju:

- **Port is open**: the workload's own `Service` is referenced directly by the
  `HTTPRoute`'s `backendRefs`.
- **Port is closed**: the Ingress Configurator creates a selector-based
  `Service` that targets the workload's pods by the
  `app.kubernetes.io/name` label. This allows routing even when the workload
  has not declared its port to Juju.

All resources created by the Ingress Configurator are labelled with
`ingress-configurator.charm.juju.is/managed-by` so they can be discovered and
cleaned up on relation departure.

## Constraints and limitations

- **Kubernetes only**: The Gateway API is a Kubernetes-native concept, so the
  `gateway-route` relation should only be used on a Kubernetes substrate.
- **HTTP backend protocol only**: The `backend-protocol` configuration option
  only accepts `http` with `gateway-route`. TLS termination happens at the
  `Gateway`, not at the backend.

> - **One route relation at a time**: The charm blocks if more than one of

Reviewer: I may be missing something: is it "One 'upstream' root at a time"? Should "gateway-route" be listed here?

Reviewer: Forget about it, I was thinking about gateway-api, not the configurator.

  `haproxy-route`, `haproxy-route-tcp`, or `gateway-route` is related
  simultaneously.

> - **One `gateway-route` relation per instance**: A single

Reviewer: To be clarified too: "per instance of ingress-configurator", right?

Reviewer: Same.

---
myst:
  html_meta:
    "description lang=en": "Explanation of how the gateway-route interface works in the Ingress Configurator charm, including the architecture, data flow, and HTTPS modes."
---

(explanation_gateway_route)=

# How gateway-route works

The `gateway-route` interface is used together with the
[`gateway-api-integrator`](https://charmhub.io/gateway-api-integrator) charm,
which manages a Kubernetes [Gateway API](https://gateway-api.sigs.k8s.io/)
`Gateway` resource. The Ingress Configurator charm connects a workload that

> only speaks the `ingress` relation to that `Gateway`, while also exposing

Reviewer 2: Nit: I find the phrasing "workload that only speaks the `ingress` relation" to be a little confusing. In particular, "speaks" is not clear what the underlying mechanism is -- are you trying to say that `ingress-configurator` only interfaces with the `Gateway` resource via the `ingress` relation?

configuration options - such as `hostname` and `paths` - that can be tuned

> without modifying either the workload or the `gateway-api-integrator`.

Reviewer 2: ```suggestion
without modifying either the workload or `gateway-api-integrator`.
```
or
```suggestion
without modifying either the workload or the `gateway-api-integrator` charm.
```

        HR["K8s: HTTPRoute(s)"]
        BS["K8s: Backend Service"]
    end

    subgraph GAI["gateway-api-integrator"]
        GW["K8s: Gateway"]
        SEC["K8s: TLS Secret"]
        DNS["K8s: DNS Record"]
    end

    GW --> HR
    HR --> BS
    BS --> WS

    Workload -- "ingress relation" --> IC
    IC -- "gateway-route relation" --> GAI

    classDef k8s stroke:#2e7d32
    classDef juju stroke:#1565c0,stroke-dasharray:5 5
    linkStyle 0,1,2,3 stroke:#2e7d32
    linkStyle 4,5 stroke:#1565c0,stroke-dasharray:5 5
```

## The gateway-route relation

The `gateway-route` interface is implemented by the
`charms.gateway_api_integrator.v1.gateway_route` library. It is a bidirectional
application databag exchange:

| Direction           | Fields                                                           | Description                                                                                                                                    |
|---------------------|------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------|

> | Requirer → Provider | `hostname`, `additional_hostnames`                               | The FQDN(s) the workload should be reachable on. Used by the provider to request TLS certificates and DNS records.                             |

Reviewer 2: Do we expect readers to know what FQDN stands for, or should we explicitly spell it out here?

Author: I think we can expect them to know what it stands for

| Provider → Requirer | `gateway_name`, `gateway_model`, `https_mode`, `gateway_address` | The identity of the `Gateway` resource so the requirer can build `HTTPRoute` `parentRefs`. `https_mode` tells the requirer how TLS is handled. |

> The `https_mode` field is one of:

Reviewer 2: ```suggestion
| Provider → Requirer | `gateway_name`, `gateway_model`, `https_mode`, `gateway_address` | The identity of the `Gateway` resource so the requirer can build `HTTPRoute` `parentRefs`. |

The `https_mode` field tells the requirer how TLS is handled. The field has three available modes:
```
Nit, I found it a little strange that `https_mode` was given more explanation in the table, so I moved it out of the table

| Mode       | Meaning                                                        |
|------------|----------------------------------------------------------------|
| `disabled` | No TLS configured; only HTTP listeners exist on the `Gateway`. |
| `enabled`  | TLS configured; both HTTP and HTTPS listeners exist.           |
| `enforced` | TLS configured and HTTP is redirected to HTTPS.                |

```{note}
A single `gateway-api-integrator` can accept multiple `gateway-route`
relations, but each `ingress-configurator` instance is limited to one
`gateway-route` relation. To route traffic for multiple workloads through the
same `Gateway`, deploy a separate `ingress-configurator` for each workload and
relate them all to the same `gateway-api-integrator`.

> ```

Reviewer 2: This is a useful note, but why does it show up here? Is this information related to `https_mode` at all?

Author: it is a note at the end of the  `The gateway-route relation` section. Maybe its position is causing this confusion. I will think of a better place

## How HTTPRoute resources are created

The Ingress Configurator builds `HTTPRoute` resources based on the
`https_mode` received from the provider and the hostname(s) from its own
configuration:

- **One HTTP route** is always created. It covers all hostnames and attaches
  to every per-hostname HTTP listener on the `Gateway` via multiple
  `parentRefs`. When no hostnames are configured, it falls back to the
  `Gateway`'s hostname-less HTTP listener.
- When `https_mode` is **`enabled`** or **`enforced`**, **one HTTPS route per
  hostname** is created, each attaching to its corresponding per-hostname HTTPS
  listener.

> - When `https_mode` is **`enforced`**, the HTTP route does **not** forward to

Reviewer 2: ```suggestion
- When `https_mode` is `enabled` or `enforced`, **one HTTPS route per
  hostname** is created, each attaching to its corresponding per-hostname HTTPS
  listener.
- When `https_mode` is `enforced`, the HTTP route does **not** forward to
```
nit, there's quite a bit of formatting in this section already, so trying to remove a bit

  the backend. Instead it issues a `301` HTTPS redirect using a
  `RequestRedirect` filter.

(explanation_backend_service_selection)=

> ## Backend Service selection

Reviewer 2: ```suggestion
## Backend service selection
```
I know that `Service` has been capitalized in the text, but it's also code-formatted. Since the header does not code-format the word, I think it should be un-capitalized

The `HTTPRoute` forwards traffic to a Kubernetes `Service`. Which `Service` is
used depends on whether the workload charm has opened its port in Juju:

- **Port is open**: the workload's own `Service` is referenced directly by the
  `HTTPRoute`'s `backendRefs`.

> - **Port is not open**: the Ingress Configurator creates a selector-based

Reviewer 2: ```suggestion
- **Port is closed**: the Ingress Configurator creates a selector-based
```
nit, to be consistent with the way ports were described previously

  `Service` that targets the workload's pods by the
  `app.kubernetes.io/name` label. This allows routing even when the workload
  has not declared its port to Juju.

All resources created by the Ingress Configurator are labelled with
`ingress-configurator.charm.juju.is/managed-by` so they can be discovered and
cleaned up on relation departure.

## Constraints and limitations

- **Kubernetes only**: the Gateway API is a Kubernetes-native concept, so the
  `gateway-route` relation should only be used on a Kubernetes substrate.
- **HTTP backend protocol only**: the `backend-protocol` configuration option
  only accepts `http` with `gateway-route`. TLS termination happens at the
  `Gateway`, not at the backend.
- **One route relation at a time**: the charm blocks if more than one of
  `haproxy-route`, `haproxy-route-tcp`, or `gateway-route` is related
  simultaneously.
- **`--trust` required**: the charm needs Kubernetes RBAC permissions to

>   manage `HTTPRoute` and `Service` resources.

Reviewer 2: ```suggestion
- **Kubernetes only**: The Gateway API is a Kubernetes-native concept, so the
  `gateway-route` relation should only be used on a Kubernetes substrate.
- **HTTP backend protocol only**: The `backend-protocol` configuration option
  only accepts `http` with `gateway-route`. TLS termination happens at the
  `Gateway`, not at the backend.
- **One route relation at a time**: The charm blocks if more than one of
  `haproxy-route`, `haproxy-route-tcp`, or `gateway-route` is related
  simultaneously.
- **`--trust` required**: The charm needs Kubernetes RBAC permissions to
  manage `HTTPRoute` and `Service` resources.
```
nit
